### PR TITLE
Smoothing Functions For Actors

### PR DESCRIPTION
**Pull request by nannick**
**Summary of changes**
Changes introduced in this pull request:
- Adds packages forest_math and forest_smooth under the utils/ directory
- Adds structs FilterEstimate and  AlphaBetaFilter
- Adds tests

**Reference issue to close (if applicable)**

Closes 
#594 

**Other information and links**

### CODE REVIEW

File: utils/math/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    for coeff in poly {
+        let temp = &res * x;
+        res = (temp >> PRECISION) + coeff;
+    }
+    res
+}
+
+pub fn parse(coefs: &[&str]) -> Result<Vec<BigInt>, ()> {
+    println!("Num elements is {}", coefs.len());
```
Comment (austinabell) [inline utils/math/src/lib.rs:19]:
```suggestion
```

File: utils/math/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    let mut out: Vec<BigInt> = Vec::with_capacity(coefs.len() as usize);
+
+    for coef in coefs {
+        let c = BigInt::parse_bytes(coef.as_bytes(), 10).ok_or(())?;
+        out.push(c);
+    }
+    Ok(out)
+}
```
Comment (austinabell) [inline utils/math/src/lib.rs:28]:
```suggestion
pub fn parse(coefs: &[&str]) -> Result<Vec<BigInt>, ParseBigIntError> {
    coefs.iter().map(|c| c.parse()).collect()
}
```
Why drop the error type or use as bytes?

File: utils/smoothing/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+forest_math = {path = "../math"}
+lazy_static = "1.4"
+num-traits = "0.2"
+actor = {path = "../../vm/actor", package = "actor"}
+fil_types = { path = "../../types" }
+encoding = { package = "forest_encoding", path = "../../encoding" }
+serde = { version = "1.0", features = ["derive"] }
+
+
+[dev-dependencies]
```
Comment (austinabell) [inline utils/smoothing/Cargo.toml:19]:
```suggestion
```

File: utils/math/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+use num_bigint::BigInt;
+
+pub const PRECISION: u64 = 128;
+
+pub fn poly_val(poly: &[BigInt], x: &BigInt) -> BigInt {
+    let mut res = BigInt::default();
+
+    for coeff in poly {
+        let temp = &res * x;
+        res = (temp >> PRECISION) + coeff;
```
Comment (austinabell) [inline utils/math/src/lib.rs:13]:
```suggestion
        res = ((res * x) >> PRECISION) + coeff;
```
No need for the temp value

File: utils/math/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,28 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use num_bigint::BigInt;
+
+pub const PRECISION: u64 = 128;
+
+pub fn poly_val(poly: &[BigInt], x: &BigInt) -> BigInt {
```
Comment (austinabell) [inline utils/math/src/lib.rs:8]: can you add docs on the public functions (unclear the exact reasoning from logic what it does)

File: utils/smoothing/src/alpha_beta_filter.rs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+use num_bigint::{bigint_ser, BigInt};
+use serde::{Deserialize, Serialize};
+
+#[derive(Default, Serialize, Deserialize)]
+pub struct FilterEstimate {
+    #[serde(with = "bigint_ser")]
+    pub pos: BigInt,
+    #[serde(with = "bigint_ser")]
+    pub velo: BigInt,
+}
```
Comment (austinabell) [inline utils/smoothing/src/alpha_beta_filter.rs:16]: This should be tuple serialized (`Serialize_tuple` and `Deserialize_tuple`)

File: utils/smoothing/src/constants.rs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    BigInt::parse_bytes(CONST_STRS[LN_2].as_bytes(), 10).unwrap()
+}
+
+pub fn get_cum_sum_ratio_epsilon() -> BigInt {
+    BigInt::parse_bytes(
+        CONST_STRS[EXTRAPOLATED_CUM_SUM_RATIO_EPSILON].as_bytes(),
+        10,
+    )
+    .unwrap()
+}
```
Comment (austinabell) [inline utils/smoothing/src/constants.rs:58]: why can't these also be lazy static? also you can just use `parse` from the `FromStr` implementation for BigInt, there is no need to parse as bytes for those

Comment (nannick) [reply utils/smoothing/src/constants.rs:58]: Thats alot better lol was following Lotus too closely

File: utils/math/src/lib.rs (hunk 2 above)
Comment (nannick) [reply utils/math/src/lib.rs]: wasnt aware of the pare function lol

File: utils/smoothing/src/lib.rs [hunk 8]
```diff
@@ -34,7 +34,7 @@ pub fn extrapolated_cum_sum_of_ratio(
 
     let squared_velo_2 = (velo_2 * velo_2) >> PRECISION;
 
-    if squared_velo_2 >= get_cum_sum_ratio_epsilon() {
+    if squared_velo_2 >= EPSILON.clone() {
```
Comment (austinabell) [inline utils/smoothing/src/lib.rs:37]: when accessing the lazy static, you don't need to clone them, can just deref and/or reference them. Should only need to clone them if you need to initialize a new variable

**Comment by ec2** [comment]
I dont think you needed to create 2 new crates for these. Can just throw them into vm/actor/src/util. I don't think these will be used anywhere else.

File: vm/actor/src/util/smooth.rs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        "340282366920938463463374607431768211456",
+    ])
+    .unwrap();
+    pub static ref LN_2: BigInt = "235865763225513294137944142764154484399".parse().unwrap();
+    pub static ref EPSILON: BigInt = "302231454903657293676544".parse().unwrap();
+}
+
+fn get_bit_len(z: &BigInt) -> usize {
+    z.abs().to_radix_le(2).1.len()
+}
```
Comment (austinabell) [inline vm/actor/src/util/smooth.rs:39]:
```suggestion
fn get_bit_len(z: &BigInt) -> u64 {
    z.bits()
}
```
Can just remove this function altogether if you'd like (unnecessary)

File: vm/actor/src/util/smooth.rs [hunk 10]
```diff
@@ -0,0 +1,101 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use num_bigint::BigInt;
+
+use super::alpha_beta_filter::*;
+
+use crate::math::{parse, poly_val, PRECISION};
```
Comment (austinabell) [inline vm/actor/src/util/smooth.rs:8]: can you remove spacing like this?

File: vm/actor/src/util/alpha_beta_filter.rs [hunk 11]
```diff
@@ -0,0 +1,71 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+#![allow(dead_code)]
+use super::math::PRECISION;
+
+use clock::ChainEpoch;
```
Comment (austinabell) [inline vm/actor/src/util/alpha_beta_filter.rs:7]: spacing

File: vm/actor/src/util/alpha_beta_filter.rs [hunk 12]
```diff
@@ -0,0 +1,71 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+#![allow(dead_code)]
```
Comment (austinabell) [inline vm/actor/src/util/alpha_beta_filter.rs:4]:
```suggestion
```

Comment (nannick) [reply vm/actor/src/util/alpha_beta_filter.rs:4]: Right now I just have this because AlphaBetaFilter is unused. I'll leave a TODO on it or something

Comment (austinabell) [reply vm/actor/src/util/alpha_beta_filter.rs:4]: it's public tho? I tried removing and got no warnings (unless something was slow)

Comment (nannick) [reply vm/actor/src/util/alpha_beta_filter.rs]: oh dude ur right. I have no idea why it was throwing that warning earlier

**Pull request merged**